safe_transaction_service/events/services/queue_service.py

```diff
@@ -0,0 +1,279 @@
+import json
+import logging
+from typing import Any, Dict, Optional
+
+from django.conf import settings
+
+import pika.exceptions
+from pika import BlockingConnection, ConnectionParameters, PlainCredentials
+from pika.adapters.gevent_connection import GeventConnection
+from pika.channel import Channel
+from pika.exchange_type import ExchangeType
+
+logger = logging.getLogger(__name__)
+
+
+class QueueServiceProvider:
+    def __new__(cls):
+        if not hasattr(cls, "instance"):
+            if settings.EVENTS_QUEUE_HOST:
+                if settings.EVENTS_QUEUE_ASYNC_CONNECTION:
+                    cls.instance = AsyncQueueService()
+                else:
+                    cls.instance = SyncQueueService()
+            else:
+                # Mock send_event to not configured host us is not mandatory configure a queue for events
+                cls.instance = MockedQueueService()
+        return cls.instance
+
+    @classmethod
+    def del_singleton(cls):
+        if hasattr(cls, "instance"):
+            del cls.instance
+
+
+class QueueService:
+    EXCHANGE_NAME = settings.EVENTS_QUEUE_RABBITMQ_EXCHANGE_NAME
+
+    def __init__(self):
+        self._channel = None
+        self._connection: GeventConnection = None
+        self.unsent_events = []
+        credentials = PlainCredentials(
+            settings.EVENTS_QUEUE_USERNAME, settings.EVENTS_QUEUE_PASSWORD
+        )
+        self._connection_parameters = ConnectionParameters(
```

Comment:
I think using URL parameters is way more flexible. I would only leave password out due to security reasons: https://pika.readthedocs.io/en/stable/modules/parameters.html#urlparameters

Reply:
Done!